File: mobile/lib/src/data/models/assignment.dart
class Assignment {
  final String id;
  final String title;
  final String description;
  final String className;
  final String section;
  final String subject;
  final String teacherId;
  final DateTime dueDate;
  final int totalMarks;
  final List<Attachment> attachments;
  final String status;
  final DateTime createdAt;

  Assignment({
    required this.id,
    required this.title,
    required this.description,
    required this.className,
    required this.section,
    required this.subject,
    required this.teacherId,
    required this.dueDate,
    required this.totalMarks,
    required this.attachments,
    required this.status,
    required this.createdAt,
  });

  bool get isOverdue => DateTime.now().isAfter(dueDate) && status == 'active';

  factory Assignment.fromJson(Map<String, dynamic> json) {
    return Assignment(
      id: json['_id'] ?? json['id'],
      title: json['title'],
      description: json['description'],
      className: json['class'],
      section: json['section'],
      subject: json['subject'],
      teacherId: json['teacherId'],
      dueDate: DateTime.parse(json['dueDate']),
      totalMarks: json['totalMarks'] ?? 100,
      attachments:
          (json['attachments'] as List?)
              ?.map((a) => Attachment.fromJson(a))
              .toList() ??
          [],
      status: json['status'],
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
}

class Attachment {
  final String filename;
  final String url;
  final int? size;

  Attachment({required this.filename, required this.url, this.size});

  factory Attachment.fromJson(Map<String, dynamic> json) {
    return Attachment(
      filename: json['filename'],
      url: json['url'],
      size: json['size'],
    );
  }

  Map<String, dynamic> toJson() {
    return {'filename': filename, 'url': url, 'size': size};
  }
}

File: server/src/models/Assignment.js
const mongoose = require('mongoose');

const assignmentSchema = new mongoose.Schema({
    title: {
        type: String,
        required: true
    },
    description: {
        type: String,
        required: true
    },
    class: {
        type: String,
        required: true,
        index: true
    },
    section: {
        type: String,
        required: true
    },
    subject: {
        type: String,
        required: true
    },
    teacherId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Teacher',
        required: true
    },
    dueDate: {
        type: Date,
        required: true,
        index: true
    },
    totalMarks: {
        type: Number,
        default: 100
    },
    attachments: [{
        filename: String,
        url: String,
        size: Number
    }],
    submissions: [{
        studentId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'Student'
        },
        submittedAt: {
            type: Date,
            default: Date.now
        },
        files: [{
            filename: String,
            url: String,
            size: Number
        }],
        marksObtained: Number,
        feedback: String,
        status: {
            type: String,
            enum: ['submitted', 'graded', 'late'],
            default: 'submitted'
        }
    }],
    status: {
        type: String,
        enum: ['active', 'closed', 'draft'],
        default: 'active'
    }
}, {
    timestamps: true
});

// Index for querying assignments
assignmentSchema.index({ class: 1, section: 1, dueDate: -1 });
assignmentSchema.index({ teacherId: 1, createdAt: -1 });

module.exports = mongoose.model('Assignment', assignmentSchema);
